lizmap/modules/lizmap/lib/Form/QgisForm.php

```diff
@@ -1031,6 +1031,22 @@ public function deleteFromDb($feature)
         return $result;
     }
 
+    public function getModifiedControls()
+    {
+        $modifiedControls = $this->getForm()->getModifiedControls();
+        $form = $this->form;
+        foreach ($this->formControls as $fieldName => $formControl) {
+            if (($this->formControls[$fieldName]->fieldEditType === 7
+            || $this->formControls[$fieldName]->fieldEditType === 'CheckBox')
+            && $this->formControls[$fieldName]->fieldDataType === 'boolean'
+            && !array_key_exists($fieldName, $modifiedControls)) {
+                $modifiedControls[$fieldName] = $this->getForm()->getControl($fieldName);
```

Comment:
You add the control into the modifiedControls list, without checking if it's state has been modified?

Reply:
yes because if checkbox is not checked the value is `null` and not `false`

Reply:
`getModifiedControls` returns controls for which value has been changed. So you should add your checkbox into $modifiedControls **only** if it's "value" has changed, so only if its state has changed (checked -> unchecked or unchecked -> checked).

And probably the issue has to be fixed into the `getModifiedControls` method. Is there a simple test case somewhere? (and/or a ticket?)

Reply:
@laurentj here is the use case:
* boolean field in datatable with not null constraint but no default value defined
* the edit widget for the field is a checkbox without default value defined
* when the form is build for creation the checkbox is unchecked
* when the form is send for creation the checkbox can be still unchecked
* in lizmap the code set the default values to get field with default values in the modified controls list, but here the field has no default value
* the checkbox will not be in the modified control list if the status checked/unchecked has not chenged but lizmap use this list to only set modified fields even if in creation and needs checkbox value even if the status has not changed.
* In this case, the creation failed because no values has been set for the boolean not null without default value with a checkbox control.

To fix the issue, we can:
* define a new method for QgisForm - rename this `getModifiedControls` by something like `getControlsForSave`
* remove this method and update `saveToDb` to get checkbox values in addition to modifiedControls

Reply:
Ok I see the point. However, when inserting a new feature (a new record), you should not use 'initModifiedControls' and 'getModifiedControls', as it is to know modified values for an **existing** record. You should use them only if `$this->featureId` has a value, into `saveFeature`. In the case of a creation, you should use all controls of the form to save data.